src/main/scala/is/hail/utils/UInt.scala

```diff
@@ -0,0 +1,78 @@
+package is.hail.utils
+
+object UInt {
+  def apply(i: Int): UInt = new UInt(i)
+
+  def apply(l: Long): UInt = {
+    assert((l >>> 32) == 0, s"High bits not set to 0. Cannot convert $l to UInt.")
+    l.toUInt
+  }
+
+  implicit val numeric: Numeric[UInt] = new Numeric[UInt] {
+    override def plus(x: UInt, y: UInt): UInt = x + y
+
+    override def minus(x: UInt, y: UInt): UInt = x - y
+
+    override def times(x: UInt, y: UInt): UInt = x * y
+
+    override def negate(x: UInt): UInt = throw new UnsupportedOperationException()
+
+    override def fromInt(x: Int): UInt = UInt(x)
+
+    override def toInt(x: UInt): Int = x.i
+
+    override def toLong(x: UInt): Long = x.toLong
+
+    override def toFloat(x: UInt): Float = x.toFloat
+
+    override def toDouble(x: UInt): Double = x.toDouble
+
+    override def compare(x: UInt, y: UInt): Int = x.compare(y)
+
+    override def zero: UInt = UInt(0)
+
+    override def one: UInt = UInt(1)
+  }
+}
+
+class UInt private(val i: Int) extends AnyVal {
+
+  def compare(right: UInt): Int = {
+    val l = toLong
+    val r = right.toLong
+    if (l == r) 0 else if (l > r) 1 else -1
+  }
+
+  def +(right: UInt): UInt = UInt(toLong + right.toLong)
+
+  def -(right: UInt): UInt = UInt(toLong - right.toLong)
+
+  def *(right: UInt): UInt = UInt(toLong * right.toLong)
+
+  def /(right: UInt): UInt = UInt(toLong / right.toLong)
+
+  def <=(right: UInt): Boolean = toLong <= right.toLong
+
+  def >=(right: UInt): Boolean = toLong >= right.toLong
+
+  def <(right: UInt): Boolean = toLong < right.toLong
+
+  def >(right: UInt): Boolean = toLong > right.toLong
+
+  def ==(right: UInt): Boolean = toLong == right.toLong
+
+  def !=(right: UInt): Boolean = toLong != right.toLong
+
+  def toInt: Int = {
+    assert(i >= 0)
```

Comment:
Ah, sorry, clearly require(i >= 0) in the class is wrong, since we can have unsigned numbers with the high bit set.  It should be tested in the UInt.apply(Int).  Right now it looks like if you construct a UInt with a negative number UInt(-2), it will work.  Am I wrong?